DEMS-Common/src/Utility/AbstractReplica.java

```diff
@@ -0,0 +1,189 @@
+package Utility;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutput;
+import java.io.ObjectOutputStream;
+import java.util.HashMap;
+
+import Interface.Corba.Project;
+import Models.AddressBook;
+import Models.RegisteredReplica;
+import UDP.Message;
+import UDP.OperationCode;
+import UDP.RequestListener;
+import UDP.RequestListener.Processor;
+
+public abstract class AbstractReplica {
+	protected RegisteredReplica replicaID;
+	protected static HashMap<String, ICenterServer> centerServers = new HashMap<>();
+	protected HashMap<Integer, Message> messagesQueue = new HashMap<>();	
+	protected int nextSequenceID = 0;
+	
+	protected class ReplicaListerner implements Processor {
+    	final private RequestListener m_Listener;
+    	private Thread m_ListenerThread;
+    	
+    	public ReplicaListerner() {
+    		m_Listener = new RequestListener(this, AddressBook.REPLICAS);
+    	}
+    	
+		public void launch(){
+			m_ListenerThread = new Thread(m_Listener);
+			m_ListenerThread.start();
+//			m_Listener.Wait(); // Make sure it's running before getting any farther ( optional )
+		}
+		
+		public void shutdown() {
+			m_Listener.Stop();
+			
+			try {
+				m_ListenerThread.join();
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
+		}
+		
+		/**
+	 	 *  Operation Handled:
+	 	 *  SERIALIZE(1011)
+	 	 */
+		@Override
+		public String handleRequestMessage(Message msg) throws Exception {
+			if (msg.getOpCode() == OperationCode.SERIALIZE && msg.getData().contains("ClientRequest")) {
+				
+				int sequenceID = msg.getSeqNum();
+				
+				if(sequenceID == nextSequenceID) {
+					nextSequenceID++;
+				}
+				else if(sequenceID > nextSequenceID) {
+					messagesQueue.put(sequenceID, msg);
+				}
+				
+				ReplicaResponse replicaResponse = AbstractReplica.this.processRequest(msg);
+				String serializedResponse = serializeReplicaResponse(replicaResponse);
+				
+				return serializedResponse;
+				
+				// Check if next request is in messagesQueue
+//				if(messagesQueue.get(nextSequenceID) != null) {
+//					handleRequestMessage(messagesQueue.get(nextSequenceID));
+//				}
```

Comment:
I don't know when to perform this check to process the requests in the queue.
It can't be here because we need to return the value for the current request.

Reply:
if you make a queue you need to queue everything and and have a separate thread... that does not work with the implementation of the RUDP server